vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,1206 @@
+                        size_of::<u8>(),
+                        writter.available_bytes() as u32,
+                    ));
+                }
+                Ordering::Equal => Vec::<u8>::new(),
+                Ordering::Greater => {
+                    if trans_len != 0
+                        && trans_len != (writter.available_bytes() - size_of::<u8>()) as u32
+                    {
+                        return Err(Error::TxRxTrnasLenNotEqual(
+                            trans_len,
+                            (writter.available_bytes() - size_of::<u8>()) as u32,
+                        ));
+                    } else if trans_len == 0 {
+                        trans_len = (writter.available_bytes() - size_of::<u8>()) as u32;
+                    }
+                    vec![0; trans_len as usize]
+                }
+            };
+
+            if trans_len == 0 {
+                return Err(Error::TransZeroLength);
+            }
+
+            let out_hdr = reader
+                .read_obj::<VirtioSpiTransferHead>()
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            reqs.push(SpiTransReq {
+                tx_buf,
+                rx_buf,
+                trans_len,
+                speed_hz: out_hdr.freq.to_native(),
+                delay_usecs: (out_hdr.cs_delay_hold_ns.to_native() / 1000) as u16,
+                bits_per_word: out_hdr.bits_per_word,
+                cs_change: out_hdr.cs_change,
+                tx_nbits: out_hdr.tx_nbits,
+                rx_nbits: out_hdr.rx_nbits,
+                word_delay_usecs: (out_hdr.word_delay_ns.to_native() / 1000) as u8,
+                mode: out_hdr.mode.to_native(),
+                cs_id: out_hdr.chip_select_id,
+            });
+        }
+
+        let in_hdr = match self.spi_ctrl.check_trans_params(&mut reqs) {
+            true => VirtioSpiTransferResult {
+                status: match self.spi_ctrl.transfer(&mut reqs) {
+                    Ok(()) => VIRTIO_SPI_TRANS_OK,
+                    Err(_) => VIRTIO_SPI_TRANS_ERR,
+                },
+            },
+            _ => VirtioSpiTransferResult {
+                status: VIRTIO_SPI_PARAM_ERR,
+            },
+        };
+
+        for (i, desc_chain) in requests.iter().enumerate() {
+            let mut len = size_of::<VirtioSpiTransferResult>() as u32;
+            let mem = atomic_mem.memory();
+
+            let mut writter = desc_chain
+                .clone()
+                .writer(&mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            if writter.available_bytes() > size_of::<u8>() {
+                let rx_len = (writter.available_bytes() - size_of::<u8>()) as u32;
+                let mut writter_status = writter
+                    .split_at(rx_len as usize)
+                    .map_err(|_| Error::DescriptorReadFailed)?;
+
+                writter_status
+                    .write_obj(in_hdr)
```

Comment:
I probably missed something, but why do you write the same status for each request? Isn't the status different for each request?

Reply:
Ok, will update the logic here